File: src/mesh/paths.ts
import type { RadiusField } from "@/lib/grower/radius";
import type { Skeleton } from "@/lib/grower/skeleton/colonize";

/* ------------------------------------------------------------------ *
 * THE SKELETON, CUT INTO RUNS
 *
 * The skeleton is a tree of nodes, and the fn-11.4 viewer drew it one
 * edge at a time: a tapered cylinder per edge, each with its own two
 * flat ends. Wherever the centreline bent, the two ends meeting at a
 * node stopped covering each other and the joint opened - the rings
 * and the see-through gaps the owner screenshotted. Nothing about that
 * is fixable at the edge level, because two independently oriented
 * discs never meet.
 *
 * So the surface is not swept over edges. It is swept over runs: a
 * continuous chain of nodes from an attachment point to a tip, with
 * ONE ring per node, shared by the segment below it and the segment
 * above it. A shared ring cannot open, because there is only one of
 * it. That is the whole of the continuity fix along a limb, and it is
 * structural rather than tuned - no epsilon, no overlap, no fudge.
 *
 * Which child continues the run and which one starts a new one is
 * decided by thickness: the thickest child is the same limb carrying
 * on, and everything else is a branch off it. That is what the radius
 * solve already means, so the surface inherits the fork rule rather
 * than forming a second opinion about which limb is the leader. Ties
 * go to the lower node index, so the decomposition is deterministic in
 * the skeleton alone.
 *
 * Every run past the first begins AT the fork node it leaves, not at
 * the child. The run therefore starts inside the parent limb's own
 * volume, which is what gives surface.ts somewhere to socket the child
 * so that a fork has no gap to show. The alternative - starting the
 * child at its own node - leaves the gap between the child's first
 * ring and the parent's surface, which is the fork half of exactly the
 * defect being fixed.
 * ------------------------------------------------------------------ */

/** Two nodes closer than this are the same point, as a multiple of the
 *  tree's own scale. Colonization can emit a step of zero length, and
 *  a zero-length segment is a ring of degenerate triangles - so those
 *  nodes are folded into their parent and their children re-parented
 *  onto it, which drops the degeneracy without dropping the subtree. */
const SAME_POINT = 1e-9;

export interface BranchPath {
  /** Node indices along one continuous run, in growth order.
   *  `nodes[0]` is where the run attaches - the skeleton's root for
   *  the trunk, the fork node for every other run. At least two long;
   *  a run with nothing to sweep is never emitted. */
  nodes: number[];
  /** True for the one run that starts at the skeleton's root and
   *  follows the thickest child at every fork: the trunk. Every other
   *  run leaves a fork and has to be socketed into its parent. */
  trunk: boolean;
}

/**
 * Cuts `skeleton` into the runs the surface is swept over.
 *
 * Pure and deterministic in the skeleton and the radius field. Every
 * edge of the skeleton appears in exactly one run, so the surface
 * covers the whole tree and covers no part of it twice.
 *
 * Returns an empty array for a skeleton that never grew - one node has
 * no run in it, and the caller gets an empty mesh rather than an
 * exception.
 */
export function branchPaths(
  skeleton: Skeleton,
  field: RadiusField,
): BranchPath[] {
  const nodes = skeleton.nodes;
  const count = nodes.length;
  if (count < 2) return [];

  /* Zero-length edges collapse onto their parent. `stands[i]` is the
     node index that node i's position is actually drawn at: itself,
     or the ancestor it is a duplicate of. Node indices rise away from
     the root, so one forward pass resolves every alias before it is
     used. */
  const stands = new Int32Array(count);
  const children: number[][] = Array.from({ length: count }, () => []);
  for (let i = 1; i < count; i += 1) {
    const parent = nodes[i].parent;
    if (parent < 0) {
      stands[i] = i;
      continue;
    }
    const at = stands[parent];
    if (nodes[i].position.distanceTo(nodes[at].position) > SAME_POINT) {
      stands[i] = i;
      children[at].push(i);
    } else {
      stands[i] = at;
    }
  }

  /** The child that carries the limb on: the thickest where it leaves
   *  the fork, ties to the lower index. */
  const leader = (of: number): number => {
    const kids = children[of];
    let best = kids[0];
    for (let k = 1; k < kids.length; k += 1) {
      if (field.startRadius[kids[k]] > field.startRadius[best]) best = kids[k];
    }
    return best;
  };

  const paths: BranchPath[] = [];
  /* Breadth-first over runs, seeded with the trunk. Each seed is the
     node the run attaches to plus the child it takes first; the side
     children found while walking a run become seeds of their own, in
     the order they were grown. */
  const seeds: { attach: number; first: number }[] = [];
  if (children[0].length > 0) seeds.push({ attach: 0, first: leader(0) });
  for (const kid of children[0]) {
    if (kid !== seeds[0]?.first) seeds.push({ attach: 0, first: kid });
  }

  for (let s = 0; s < seeds.length; s += 1) {
    const seed = seeds[s];
    const run = [seed.attach, seed.first];
    let at = seed.first;
    while (children[at].length > 0) {
      const next = leader(at);
      for (const kid of children[at]) {
        if (kid !== next) seeds.push({ attach: at, first: kid });
      }
      run.push(next);
      at = next;
    }
    paths.push({ nodes: run, trunk: s === 0 });
  }

  return paths;
}

File: crates/telperion-core/src/surface/paths.rs
use super::{filled, reserved};
use crate::{math::Vec3, tree::Node, Error, Result};
pub(super) struct Run {
    pub start: usize,
    pub end: usize,
    pub trunk: bool,
}
pub(super) struct Paths {
    pub nodes: Vec<usize>,
    pub runs: Vec<Run>,
    /// Every node where a clump's stems part above the ground, and the run
    /// through it carries on into the straighter of them.
    pub forks: Vec<bool>,
}
pub(super) fn paths(nodes: &[Node]) -> Result<Paths> {
    let n = nodes.len();
    let mut result = Paths {
        nodes: reserved(
            n.checked_mul(2)
                .ok_or(Error::ResourceLimit("surface paths"))?,
        )?,
        runs: reserved(n)?,
        forks: filled(n, false)?,
    };
    if n < 2 {
        return Ok(result);
    }
    let mut stands = filled(n, 0)?;
    let mut first = filled(n, usize::MAX)?;
    let mut last = filled(n, usize::MAX)?;
    let mut next = filled(n, usize::MAX)?;
    let mut leaders = filled(n, usize::MAX)?;
    for i in 1..n {
        let at = stands[nodes[i].parent.unwrap() as usize];
        if nodes[i].position.distance(nodes[at].position) > 1e-9 {
            stands[i] = i;
            if first[at] == usize::MAX {
                first[at] = i;
            } else {
                next[last[at]] = i;
            }
            last[at] = i;
            if leaders[at] == usize::MAX || nodes[i].start_radius > nodes[leaders[at]].start_radius
            {
                leaders[at] = i;
            }
        } else {
            stands[i] = at;
        }
    }
    for at in (1..n).filter(|&at| stands[at] == at && nodes[at].stem) {
        let before = nodes[stands[nodes[at].parent.unwrap() as usize]].position;
        let child = |k: usize| Some(k).filter(|&k| k != usize::MAX);
        let stems = std::iter::successors(child(first[at]), |&k| child(next[k]))
            .filter(|&k| nodes[k].stem)
            .map(|k| (k, nodes[k].position));
        if let Some(leader) = straightest(before, nodes[at].position, stems) {
            leaders[at] = leader;
            result.forks[at] = true;
        }
    }
    let mut seeds = reserved(n)?;
    if leaders[0] != usize::MAX {
        seeds.push((0, leaders[0]));
    }
    let mut k = first[0];
    while k != usize::MAX {
        if k != leaders[0] {
            seeds.push((0, k));
        }
        k = next[k];
    }
    let mut s = 0;
    while s < seeds.len() {
        let (attach, mut at) = seeds[s];
        let start = result.nodes.len();
        result.nodes.extend([attach, at]);
        while leaders[at] != usize::MAX {
            let leader = leaders[at];
            let mut k = first[at];
            while k != usize::MAX {
                if k != leader {
                    seeds.push((at, k));
                }
                k = next[k];
            }
            result.nodes.push(leader);
            at = leader;
        }
        result.runs.push(Run {
            start,
            end: result.nodes.len(),
            // Every run that leaves the root is a trunk run: a buried root
            // sample, the flare its own height earns it, and no fork socket
            // where it meets the ground. A tree on one stem has exactly one,
            // which is the run that was the trunk before stems were a row.
            // A stem a clump parts from its first above the ground leaves
            // wood and not the ground, so it is swept from the socket of the
            // trunk below the fork like any run that forks off, unburied.
            trunk: attach == 0,
        });
        s += 1;
    }
    Ok(result)
}

/// Where two or more stems leave a node on a stem above the root, the run
/// carries on into the one that turns least from the wood below it, the
/// first of them on a tie, and every other leaves it as a socketed run. Any
/// fewer stems and the run follows its widest child, as everywhere else.
/// `before` is where the wood below the node starts, `at` the node, and
/// `stems` its stem children in child order with their positions. The
/// contact query asks the same of the same positions, so both choose one run.
pub(crate) fn straightest<K>(
    before: Vec3,
    at: Vec3,
    stems: impl IntoIterator<Item = (K, Vec3)>,
) -> Option<K> {
    let mut stems = stems.into_iter();
    let below = (at - before).normalized();
    let turn = |p: Vec3| below.dot((p - at).normalized());
    let (mut best, p) = stems.next()?;
    let mut most = turn(p);
    let mut parted = false;
    for (k, p) in stems {
        parted = true;
        let straightness = turn(p);
        if straightness > most {
            (best, most) = (k, straightness);
        }
    }
    parted.then_some(best)
}
